crates/polars-stream/src/utils/byte_source.rs

```diff
@@ -0,0 +1,329 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use polars_error::{to_compute_err, PolarsResult};
+use polars_io::cloud::{
+    build_object_store, object_path_from_str, CloudLocation, CloudOptions, ObjectStorePath,
+    PolarsObjectStore,
+};
+use polars_utils::{_limit_path_len_io_err, no_call_const};
+
+use crate::async_executor;
+
+mod bytes_private {
+    use std::sync::Arc;
+
+    /// This is for keeping the buffers alive.
+    #[derive(Clone)]
+    #[allow(unused)]
+    pub(super) enum BytesKeepAlive {
+        Local(Arc<memmap::Mmap>),
+        Cloud(bytes::Bytes),
+    }
+
+    /// TODO: Add slicing to owned data
+    #[derive(Clone)]
+    pub struct Bytes {
+        // Store the `&[u8]` to make `AsRef<[u8]>` free.
+        // `slice` is not 'static - it is backed by `keep_alive`. This is safe as long as `slice` is
+        // not directly accessed, and we are in a private module to guarantee that. Access should
+        // only be done through `AsRef<[u8]>`, which automatically gives the correct lifetime.
+        slice: &'static [u8],
+        _keep_alive: BytesKeepAlive,
+    }
+
+    impl Bytes {
+        pub(super) fn new(slice: &'static [u8], _keep_alive: BytesKeepAlive) -> Self {
+            Self { slice, _keep_alive }
+        }
+    }
+
+    impl AsRef<[u8]> for Bytes {
+        fn as_ref(&self) -> &[u8] {
+            self.slice
+        }
+    }
+}
+
+pub use bytes_private::Bytes;
+use bytes_private::BytesKeepAlive;
+
+pub trait ByteSource: Sized + Send + Sync {
+    async fn try_new_from_path(
+        path: &str,
+        cloud_options: Option<&CloudOptions>,
+    ) -> PolarsResult<Self>;
+    async fn get_size(&self) -> PolarsResult<usize>;
+    async fn get_range(&self, range: Range<usize>) -> PolarsResult<Bytes>;
+}
+
+pub struct LocalByteSource {
+    mmap: Arc<memmap::Mmap>,
+    _file: Arc<std::fs::File>,
+}
+
+impl ByteSource for LocalByteSource {
+    async fn try_new_from_path(
+        path: &str,
+        _cloud_options: Option<&CloudOptions>,
+    ) -> PolarsResult<Self> {
+        let file = Arc::new(
+            tokio::fs::File::open(path)
+                .await
+                .map_err(|err| _limit_path_len_io_err(path.as_ref(), err))?
+                .into_std()
+                .await,
+        );
+        let mmap = Arc::new(unsafe { memmap::Mmap::map(file.as_ref()) }.map_err(to_compute_err)?);
+
+        Ok(Self { mmap, _file: file })
+    }
+
+    async fn get_size(&self) -> PolarsResult<usize> {
+        Ok(self.mmap.as_ref().len())
+    }
+
+    async fn get_range(&self, range: Range<usize>) -> PolarsResult<Bytes> {
+        // TODO: Change from mmap to file-read - this currently doesn't trigger any reads.
+        let slice = unsafe { std::mem::transmute(self.mmap.as_ref().as_ref().get(range).unwrap()) };
+
+        Ok(Bytes::new(slice, BytesKeepAlive::Local(self.mmap.clone())))
```

Comment:
This should probably use the parquet::MemReader or parquet::MemReaderSlice to ensure that it is heap-allocated and can be passed to the parquet reader.